<a href="https://colab.research.google.com/github/JF11579/Adelphi/blob/main/Colab_4_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#How did we select our Models?

We chose LAssobecasue it is reccommended  for situtationsa where there are more Features than Observations.  We used Grid Serach on Lasso to optimize it sparameters.

Lasso's is a linear regression technique that methodically reduces the Feature with the lowest coeficiaent then it re-runs the regression and again eliminstes the lowest.  

The combinations are vast.  And what it we reduce each variable but a differnt value.  Thisis what Grid Search does

We also went with ..............

We used Lasso with Grid Search to select the Features

'Seasonal', 'Gray_cement_dispatch_comercial', 'ISE',
       'Construction_lic_area/unt_SIlag_4', 'IEClag_5', 'ISElag_6',
       'Gray_cement_dispatch_Factorylag_9', 'ICElag_9',
       'Housing_total_sales', 'Housing_total_initiation_SIlag_24',
       'Ordinary_creditlag_4', 'Ordinary_creditlag_6', 'working_day']

Here is the abridged version of the code

First time through we use LAsso and Grid Search to identify our Features

## Libraries

In [28]:
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV
import pandas as pd
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.linear_model import LogisticRegression
import datetime as dt
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import Normalizer
from sklearn import preprocessing 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import Lasso
from sklearn.linear_model import Lasso, LogisticRegression
from sklearn.feature_selection import SelectFromModel
import plotly.express as px 
from numpy import mean
from numpy import std
from numpy import absolute
from pandas import read_csv
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedKFold
from numpy import mean
from numpy import std
from numpy import absolute
from pandas import read_csv
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedKFold
from sklearn.linear_model import Lasso


## Data

In [59]:
data= pd.read_csv('/content/drive/MyDrive/Porcelana_OP_adelphi_2021_SCALED.csv')
data.head(2)

date  Quantity  Seasonal  ...  holiday  sunday  working_day
0  2017-01-01  1.503548  0.699923  ...        1       5           25
1  2017-02-01  0.076213  0.316156  ...        0       4           24

[2 rows x 943 columns]

## Dating

What happends on Monday has some influence on what happens onb Tuesday.  Even if we are not going to scale the date column we cannot process the data set as is .  Weneed to change the format tof  the data column. W There are a few ways of doing this including converting them into arbtriry neumbers.  We chose to convert our dates to ORDIANL.  this maintains some relation between the numbers.  In the future we will explore Face boodk's Prophet appraoch to dates.

In [60]:
data['date'] = pd.to_datetime(data['date'])
data['date']= data['date'].map(dt.datetime.toordinal)

Later we will need this list of Feature names 

In [61]:
feature_list = data.loc[: ,data.columns != 'Quantity']
features = list(feature_list)
data.head(3)

# Notice the date column

date  Quantity  Seasonal  ...  holiday  sunday  working_day
0  736330  1.503548  0.699923  ...        1       5           25
1  736361  0.076213  0.316156  ...        0       4           24
2  736389 -1.346258 -0.267695  ...        1       4           26

[3 rows x 943 columns]

Last round we peeled off  the data column , scaled the data, then re-attached the date column.  Might have guessed that Python had an esier way to approach this.

https://stackoverflow.com/questions/62997793/apply-standarization-to-all-columns-except-one-in-pandas

In [81]:
# except B, take all columns
#cols = df.columns.difference(['B'])
#
cols =data.columns.difference(['date'])

In [82]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

In [83]:
# normalise only selected columns
#df[cols] = sc.fit_transform(df[cols])
#
data[cols] = sc.fit_transform(data[cols])

The date column is Ordial while the reso of te 943 Features are scaled  

In [84]:
data.head(3)

date  Quantity  Seasonal  ...   holiday    sunday  working_day
0  736330  1.539880  0.703768  ... -0.394158  1.357242     0.380225
1  736361  0.110421  0.325418  ... -0.653725 -0.736788     0.120324
2  736389 -1.314169 -0.250193  ... -0.394158 -0.736788     0.640126

[3 rows x 943 columns]

## Split the data into Training and Testing

y is our dependent variable/Feature

In [85]:
y = data['Quantity']

 X is every other Feature EXCEPT Quantity

In [86]:
X = data.loc[: ,data.columns != 'Quantity']

In [87]:
model = Lasso()

In [88]:
model.fit(X_train, y_train)

Lasso()

In [89]:
print("Accuracy score for training data", model.score(X_train,y_train))

Accuracy score for training data 0.06614793513532946


In [90]:
print("Accuracy score for test Data", model.score(X_test,y_test))

Accuracy score for test Data 0.043636967663260195


## Grid Search

In [91]:
pipeline = Pipeline([
                     ('scaler',StandardScaler()),
                     ('model',Lasso())
])

"search" will iterate Lasso's "Alpha" from 0.1 to 10 in 0.1 incriments

In [92]:
search = GridSearchCV(pipeline, # <<<--- "pipline" = StandardScaler and  Lasso regression
                      {'model__alpha':np.arange(0.1,10,0.1)}, # Alpha setting  
                      cv = 5, scoring="neg_mean_squared_error",verbose=3
                      )

Nwo we call Search 

In [93]:
search # Spell out the Alpha setting iterations 0.1 0.2...9.9

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model', Lasso())]),
             param_grid={'model__alpha': array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. , 1.1, 1.2, 1.3,
       1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2. , 2.1, 2.2, 2.3, 2.4, 2.5, 2.6,
       2.7, 2.8, 2.9, 3. , 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9,
       4. , 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5. , 5.1, 5.2,
       5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6. , 6.1, 6.2, 6.3, 6.4, 6.5,
       6.6, 6.7, 6.8, 6.9, 7. , 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8,
       7.9, 8. , 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9. , 9.1,
       9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9])},
             scoring='neg_mean_squared_error', verbose=3)

In [94]:
search.fit(X_train,y_train)

Fitting 5 folds for each of 99 candidates, totalling 495 fits
[CV 1/5] END .................model__alpha=0.1;, score=-0.607 total time=   0.0s
[CV 2/5] END .................model__alpha=0.1;, score=-0.725 total time=   0.1s
[CV 3/5] END .................model__alpha=0.1;, score=-0.330 total time=   0.1s
[CV 4/5] END .................model__alpha=0.1;, score=-0.655 total time=   0.1s
[CV 5/5] END .................model__alpha=0.1;, score=-0.541 total time=   0.1s
[CV 1/5] END .................model__alpha=0.2;, score=-0.487 total time=   0.0s
[CV 2/5] END .................model__alpha=0.2;, score=-0.634 total time=   0.1s
[CV 3/5] END .................model__alpha=0.2;, score=-0.162 total time=   0.1s
[CV 4/5] END .................model__alpha=0.2;, score=-0.780 total time=   0.1s
[CV 5/5] END .................model__alpha=0.2;, score=-0.433 total time=   0.1s
[CV 1/5] END .model__alpha=0.30000000000000004;, score=-0.440 total time=   0.0s
[CV 2/5] END .model__alpha=0.30000000000000004;

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model', Lasso())]),
             param_grid={'model__alpha': array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. , 1.1, 1.2, 1.3,
       1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2. , 2.1, 2.2, 2.3, 2.4, 2.5, 2.6,
       2.7, 2.8, 2.9, 3. , 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9,
       4. , 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5. , 5.1, 5.2,
       5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6. , 6.1, 6.2, 6.3, 6.4, 6.5,
       6.6, 6.7, 6.8, 6.9, 7. , 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8,
       7.9, 8. , 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9. , 9.1,
       9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9])},
             scoring='neg_mean_squared_error', verbose=3)

In [71]:
X_train, X_test,y_train,y_test = train_test_split(X , y , test_size= 0.5 , random_state= 1)

In [72]:
model.fit(X_train, y_train)

Lasso()

In [73]:
search.fit(X_train,y_train)

Fitting 5 folds for each of 99 candidates, totalling 495 fits
[CV 1/5] END .................model__alpha=0.1;, score=-0.607 total time=   0.0s
[CV 2/5] END .................model__alpha=0.1;, score=-0.725 total time=   0.1s
[CV 3/5] END .................model__alpha=0.1;, score=-0.330 total time=   0.1s
[CV 4/5] END .................model__alpha=0.1;, score=-0.655 total time=   0.1s
[CV 5/5] END .................model__alpha=0.1;, score=-0.541 total time=   0.1s
[CV 1/5] END .................model__alpha=0.2;, score=-0.487 total time=   0.1s
[CV 2/5] END .................model__alpha=0.2;, score=-0.634 total time=   0.1s
[CV 3/5] END .................model__alpha=0.2;, score=-0.162 total time=   0.1s
[CV 4/5] END .................model__alpha=0.2;, score=-0.780 total time=   0.1s
[CV 5/5] END .................model__alpha=0.2;, score=-0.433 total time=   0.0s
[CV 1/5] END .model__alpha=0.30000000000000004;, score=-0.440 total time=   0.1s
[CV 2/5] END .model__alpha=0.30000000000000004;

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model', Lasso())]),
             param_grid={'model__alpha': array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. , 1.1, 1.2, 1.3,
       1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2. , 2.1, 2.2, 2.3, 2.4, 2.5, 2.6,
       2.7, 2.8, 2.9, 3. , 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9,
       4. , 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5. , 5.1, 5.2,
       5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6. , 6.1, 6.2, 6.3, 6.4, 6.5,
       6.6, 6.7, 6.8, 6.9, 7. , 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8,
       7.9, 8. , 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9. , 9.1,
       9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9])},
             scoring='neg_mean_squared_error', verbose=3)

Lasso's Alpha works by multiplying each Feature value in turn .  the goeal is to get as many Features coefficients to zero as it can . Something x 0 = nothing and that is how Leduces dimesionality .  Counter intuitivly multiplying the Features by less than the bigest number in this case was best

In [95]:
search.best_params_

{'model__alpha': 0.4}

In [96]:
coefficients = search.best_estimator_.named_steps['model'].coef_

In [ ]:
coefficients

In [98]:
importance = np.abs(coefficients)

In [99]:
importance

array([0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.     

In [100]:
np.array(features)[importance > 0]

array(['RADAR_Plumbinglag_6', 'Housing_total_sales_SIlag_20'],
      dtype='<U37')

In [80]:
np.array(features)[importance != 0]

array(['RADAR_Plumbinglag_6', 'Housing_total_sales_SIlag_20'],
      dtype='<U37')

side by side

In [101]:
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import Lasso

In [102]:
data= pd.read_csv('/content/drive/MyDrive/Porcelana_OP_adelphi_2021_SCALED.csv')
data.head(2)

date  Quantity  Seasonal  ...  holiday  sunday  working_day
0  2017-01-01  1.503548  0.699923  ...        1       5           25
1  2017-02-01  0.076213  0.316156  ...        0       4           24

[2 rows x 943 columns]

In [103]:
data['date'] = pd.to_datetime(data['date'])
data['date']= data['date'].map(dt.datetime.toordinal)

In [104]:
#y = data_robust_scaled_named['diagnosis']
y = data['Quantity']

In [105]:
X = data.loc[: ,data.columns != 'Quantity']

In [106]:
#X = data.loc[: ,data.columns != 'Quantity']
feature_list = data.loc[: ,data.columns != 'Quantity']
#features = list(data)

In [107]:
features = list(feature_list)

In [108]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [109]:
pipeline = Pipeline([
                     ('scaler',StandardScaler()),
                     ('model',Lasso())
])

In [110]:
search = GridSearchCV(pipeline,
                      {'model__alpha':np.arange(0.1,10,0.1)},
                      cv = 5, scoring="neg_mean_squared_error",verbose=3
                      )

In [111]:
search.fit(X_train,y_train)

Fitting 5 folds for each of 99 candidates, totalling 495 fits
[CV 1/5] END .................model__alpha=0.1;, score=-0.483 total time=   0.1s
[CV 2/5] END .................model__alpha=0.1;, score=-0.723 total time=   0.1s
[CV 3/5] END .................model__alpha=0.1;, score=-1.598 total time=   0.1s
[CV 4/5] END .................model__alpha=0.1;, score=-0.668 total time=   0.1s
[CV 5/5] END .................model__alpha=0.1;, score=-0.729 total time=   0.1s
[CV 1/5] END .................model__alpha=0.2;, score=-0.383 total time=   0.1s
[CV 2/5] END .................model__alpha=0.2;, score=-0.774 total time=   0.1s
[CV 3/5] END .................model__alpha=0.2;, score=-1.796 total time=   0.1s
[CV 4/5] END .................model__alpha=0.2;, score=-0.724 total time=   0.1s
[CV 5/5] END .................model__alpha=0.2;, score=-0.478 total time=   0.1s
[CV 1/5] END .model__alpha=0.30000000000000004;, score=-0.432 total time=   0.1s
[CV 2/5] END .model__alpha=0.30000000000000004;

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model', Lasso())]),
             param_grid={'model__alpha': array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. , 1.1, 1.2, 1.3,
       1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2. , 2.1, 2.2, 2.3, 2.4, 2.5, 2.6,
       2.7, 2.8, 2.9, 3. , 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9,
       4. , 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5. , 5.1, 5.2,
       5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6. , 6.1, 6.2, 6.3, 6.4, 6.5,
       6.6, 6.7, 6.8, 6.9, 7. , 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8,
       7.9, 8. , 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9. , 9.1,
       9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9])},
             scoring='neg_mean_squared_error', verbose=3)

In [112]:
search.best_params_

{'model__alpha': 0.2}

In [113]:
coefficients = search.best_estimator_.named_steps['model'].coef_

In [114]:
importance = np.abs(coefficients)

In [ ]:
#importance

In [115]:
np.array(features)[importance > 0]

array(['Seasonal', 'Gray_cement_dispatch_comercial', 'ISE',
       'Construction_lic_area/unt_SIlag_4', 'IEClag_5', 'ISElag_6',
       'Gray_cement_dispatch_Factorylag_9', 'ICElag_9',
       'Housing_total_sales', 'Housing_total_initiation_SIlag_24',
       'Ordinary_creditlag_4', 'Ordinary_creditlag_6', 'working_day'],
      dtype='<U37')

In [118]:
Selected_Features = ['Seasonal', 'Gray_cement_dispatch_comercial', 'ISE',
       'Construction_lic_area/unt_SIlag_4', 'IEClag_5', 'ISElag_6',
       'Gray_cement_dispatch_Factorylag_9', 'ICElag_9',
       'Housing_total_sales', 'Housing_total_initiation_SIlag_24',
       'Ordinary_creditlag_4', 'Ordinary_creditlag_6', 'working_day']

In [119]:
Selected_Features

['Seasonal',
 'Gray_cement_dispatch_comercial',
 'ISE',
 'Construction_lic_area/unt_SIlag_4',
 'IEClag_5',
 'ISElag_6',
 'Gray_cement_dispatch_Factorylag_9',
 'ICElag_9',
 'Housing_total_sales',
 'Housing_total_initiation_SIlag_24',
 'Ordinary_creditlag_4',
 'Ordinary_creditlag_6',
 'working_day']